### ISSUE TITLE
Bad self type inference with Generics and inheritance?

### ISSUE THREAD

**Issue by Phlogistique** [python/mypy#9322]
* Are you reporting a bug, or opening a feature request?

Reporting a bug

* Please insert below the code you are checking with mypy

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class Foo(Generic[T]):
    def __init__(self): ...

class Bar(Generic[T]): ...

class FooChild(Foo[Bar[T]]):
    def __init__(self, blah: T):
        Foo.__init__(self)
```

* What is the actual behavior/output?

```
$ mypy foo.py
foo.py:18: error: Argument 1 to "__init__" has incompatible type "FooChild[T]"; expected "Foo[T]"
Found 1 error in 1 file (checked 1 source file)
```

* What is the behavior/output you expect?

```
Success: no issues found in 1 source file
```

* What are the versions of mypy and Python you are using?

```
mypy 0.750
Python 3.7.3
```

* Do you see the same issue after installing mypy from Git master?

Yes, successfully reproduced with `mypy 0.790+dev.df6894b5a9bb1a913e66df87138cd7a240c09fc1`

* What are the mypy flags you are using? (For example --strict-optional)

No flags.

* If mypy crashed with a traceback, please paste
  the full traceback below.

No crash

* Additional information:

The following versions typecheck successfully:

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class Foo(Generic[T]):
    def __init__(self): ...

class Bar(Generic[T]): ...

class FooChild(Foo[T]): # Removed one level of nesting here
    def __init__(self, blah: T):
        Foo.__init__(self)
```

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class Foo(Generic[T]):
    def __init__(self): ...

class Bar(Generic[T]): ...

class FooChild(Foo[Bar[T]]):
    def __init__(self): # Removed the argument here
        Foo.__init__(self)
```

**Comment by JelleZijlstra** [comment]
Maybe mypy should still support this pattern, but the more idiomatic way to write this would be `super().__init__()`.

**Comment by Phlogistique** [comment]
Indeed, the code type checks properly when using either `super().__init__()` or `super(FooChild, self).__init__()`.

My actual use case involves passing different arguments to multiple base classes' `__init__`, thus the need for naming the base class.
